### PR TITLE
google-vertexai: added langchain_google_vertexai package

### PR DESCRIPTION
**Pull request by lkuligin**
added langchain_google_vertexai package

### CODE REVIEW

File: libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    is_codey_model,
+    is_gemini_model,
+)
+from langchain_google_vertexai.utils import (
+    load_image_from_gcs,
+    raise_vertex_import_error,
+)
+
+if TYPE_CHECKING:
+    from vertexai.language_models import (
```
Comment (hwchase17) [inline libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py:41]: if we want, we can make this required now?

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+
+from langchain_google_vertexai.utils import (
+    create_retry_decorator,
+    get_client_info,
+    init_vertexai,
+    raise_vertex_import_error,
+)
+
+if TYPE_CHECKING:
+    from google.cloud.aiplatform.gapic import (
```
Comment (hwchase17) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:22]: if we want, we can make this required now

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+    )
+    from google.cloud.aiplatform.models import Prediction
+    from google.protobuf.struct_pb2 import Value
+    from vertexai.language_models._language_models import (
+        TextGenerationResponse,
+        _LanguageModel,
+    )
+    from vertexai.preview.generative_models import Image
+
+# This is for backwards compatibility
```
Comment (hwchase17) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:34]: we can remove this

File: libs/partners/google-vertexai/langchain_google_vertexai/utils.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+from typing import TYPE_CHECKING, Any, Callable, Optional, Union
+
+from langchain_core.callbacks import (
+    AsyncCallbackManagerForLLMRun,
+    CallbackManagerForLLMRun,
+)
+from langchain_core.language_models.llms import BaseLLM, create_base_retry_decorator
+
+if TYPE_CHECKING:
+    from google.api_core.gapic_v1.client_info import ClientInfo
```
Comment (hwchase17) [inline libs/partners/google-vertexai/langchain_google_vertexai/utils.py:12]: can make required if we want

File: libs/partners/google-vertexai/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+name = "langchain-google-vertexai"
+version = "0.0.1"
+description = "An integration package connecting GoogleVertexAI and LangChain"
+authors = []
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<4.0"
+langchain-core = ">=0.0.12"
+google-cloud-aiplatform = "1.38.1"
```
Comment (hwchase17) [inline libs/partners/google-vertexai/pyproject.toml:11]: is this different than the `vertexai` imports? do we want to make that one required as well?

File: libs/partners/google-vertexai/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,89 @@
+[tool.poetry]
+name = "langchain-google-vertexai"
+version = "0.0.1"
+description = "An integration package connecting GoogleVertexAI and LangChain"
+authors = []
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<4.0"
+langchain-core = ">=0.0.12"
```
Comment (hwchase17) [inline libs/partners/google-vertexai/pyproject.toml:10]: @efriis should we do `>=0.1, <0.2` now?

File: libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py (hunk 1 above)
Comment (lkuligin) [reply libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py]: done

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py (hunk 2 above)
Comment (lkuligin) [reply libs/partners/google-vertexai/langchain_google_vertexai/llms.py]: done

File: libs/partners/google-vertexai/langchain_google_vertexai/utils.py (hunk 4 above)
Comment (lkuligin) [reply libs/partners/google-vertexai/langchain_google_vertexai/utils.py]: Done

File: libs/partners/google-vertexai/pyproject.toml (hunk 5 above)
Comment (lkuligin) [reply libs/partners/google-vertexai/pyproject.toml:11]: `vertexai` is part of `google-cloud-aiplatform`

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+def _completion_with_retry(
+    llm: VertexAI,
+    prompt: List[Union[str, Image]],
+    stream: bool = False,
+    is_gemini: bool = False,
+    run_manager: Optional[CallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the completion call."""
+    retry_decorator = create_retry_decorator(run_manager=run_manager)
```
Comment (efriis) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:54]: I think we might need to plumb `max_retries` through here in order to make it retry, unless there's a way to configure retries in the latest vertexai sdk?

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+
+async def _acompletion_with_retry(
+    llm: VertexAI,
+    prompt: str,
+    is_gemini: bool = False,
+    run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the completion call."""
+    retry_decorator = create_retry_decorator(run_manager=run_manager)
```
Comment (efriis) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:80]: same here

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+class _VertexAIBase(BaseModel):
+    project: Optional[str] = None
+    "The default GCP project to use when making Vertex API calls."
+    location: str = "us-central1"
+    "The default location to use when making API calls."
+    request_parallelism: int = 5
+    "The amount of parallelism allowed for requests issued to VertexAI models. "
+    "Default is 5."
+    max_retries: int = 6
+    """The maximum number of retries to make when generating."""
```
Comment (efriis) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:104]: not used currently

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+    stop: Optional[List[str]] = None
+    "Optional list of stop words to use when generating."
+    model_name: Optional[str] = None
+    "Underlying model name."
+
+    @classmethod
+    def _get_task_executor(cls, request_parallelism: int = 5) -> Executor:
+        if cls.task_executor is None:
+            cls.task_executor = ThreadPoolExecutor(max_workers=request_parallelism)
+        return cls.task_executor
```
Comment (efriis) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:115]: is this used / can we remove?

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+    project: Optional[str] = None
+    "The default GCP project to use when making Vertex API calls."
+    location: str = "us-central1"
+    "The default location to use when making API calls."
+    request_parallelism: int = 5
+    "The amount of parallelism allowed for requests issued to VertexAI models. "
+    "Default is 5."
+    max_retries: int = 6
+    """The maximum number of retries to make when generating."""
+    task_executor: ClassVar[Optional[Executor]] = Field(default=None, exclude=True)
```
Comment (efriis) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:105]: don't have to do now to keep migration straightforward, but could be good to move this to only be in embeddings if it's only used there in 0.2

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py (hunk 9 above)
Comment (lkuligin) [reply libs/partners/google-vertexai/langchain_google_vertexai/llms.py:108]: fixed

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py (hunk 8 above)
Comment (lkuligin) [reply libs/partners/google-vertexai/langchain_google_vertexai/llms.py]: done

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py (hunk 7 above)
Comment (lkuligin) [reply libs/partners/google-vertexai/langchain_google_vertexai/llms.py]: done

File: libs/partners/google-vertexai/README.md
[the remark this answers is not in the document]
Comment (lkuligin) [reply libs/partners/google-vertexai/README.md:1]: added an initial version, we'll adjust it later then.

**Review by efriis — APPROVED** [review]
Looks good! Thanks

**Pull request merged**